### PR TITLE
refactor(`iroh-api`): add takes a `UnixfsEntry` not a `Path`

### LINKED ISSUE #458: iroh API methods should not write or read to disk
An API call currently calls `save_get_stream` which writes to disk:

https://github.com/n0-computer/iroh/blob/bd68e487ed0d0e07806f36eea4f4d8b5edcdef85/iroh-api/src/api_ext.rs#L13-L33

This shouldn't happen. Instead we should surface a helper method that makes it easy to write to disk, but the result should be streamed back as an API response that the user has more options to deal with.

Our API reads & writes to disk in many contexts. We should audit & remove these for _all_ of the iroh API

### PR DESCRIPTION
**Pull request by ramfox**
closes #458 

- [x] re-export builder structs from `iroh-unixfs::builder`
- [x] allow building from a path in `DirectoryBuilder`
- [x] make `builder::Entry` public
- [x] refactor `add` to take an `Entry`
- [x] remove unneeded `add_file`, `add_symlink`, `add_dir` helper functions
- [x] adjust imports/tests/benches

### CODE REVIEW

File: iroh-api/benches/add.rs [hunk 1]
```diff
@@ -42,30 +45,40 @@ pub fn add_benchmark(c: &mut Criterion) {
...
+                        },
+                    )
+                    .unwrap();
+                    let resolver = Resolver::new(content_loader);
+
+                    (task, client, resolver)
                 });
                 b.to_async(&executor).iter(|| {
-                    let rpc = rpc.clone();
+                    let api = Api::from_client_and_resolver(client.clone(), resolver.clone());
```
Comment (ramfox) [inline iroh-api/benches/add.rs:70]: I tried this first:
```rust
let (_task, api) = executor.block_on(async {
   ...
   let api = Api::from_config(api_config).await.unwrap();
   (task, api)
});
b.to_async(&executor).iter(|| {
   let api = api.clone();
});
```
But the complier said "no clone found in this scope", even those I derived `Clone` on `Api`, and tried implementing it manually.

Any thoughts on why I couldn't do this?

Comment (dignifiedquire) [reply iroh-api/benches/add.rs:71]: you might need to do this

```rust
```

Comment (dignifiedquire) [reply iroh-api/benches/add.rs:71]: also make sure you are not running into the mocks

File: iroh-api/benches/add.rs [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
 use iroh_unixfs::{
-    builder::Config as UnixfsConfig,
     chunker::{ChunkerConfig, DEFAULT_CHUNKS_SIZE},
+    content_loader::{FullLoader, FullLoaderConfig},
 };
 use tokio::runtime::Runtime;
 
+use iroh_api::{Api, UnixfsConfig, UnixfsEntry};
+
 pub fn add_benchmark(c: &mut Criterion) {
```
Comment (dignifiedquire) [inline iroh-api/benches/add.rs:16]: pretty sure this doesn't need to be `pub`

File: iroh-api/src/block_store.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    async fn put_many(&self, blocks: Vec<Block>) -> Result<()> {
+        let mut this = self.lock().await;
+        for block in blocks {
+            this.insert(*block.cid(), block.data().clone());
+        }
+        Ok(())
+    }
+}
+
+use async_stream::stream;
```
Comment (dignifiedquire) [inline iroh-api/src/block_store.rs:57]: please move to the top

File: iroh-api/src/block_store.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                cid,
+                raw_data_size,
+            ));
+        }
+        // make sure to also send the last chunk!
+        store.put_many(chunk).await?;
+    }
+}
+
+fn _add_blocks_to_store_single<S: Store>(
```
Comment (dignifiedquire) [inline iroh-api/src/block_store.rs:89]: do we need to keep this around?

File: iroh-unixfs/src/builder.rs [hunk 5]
```diff
@@ -411,6 +452,8 @@ impl Default for DirectoryBuilder {
             name: None,
             entries: Default::default(),
             typ: DirectoryType::Basic,
+            chunker: None,
```
Comment (dignifiedquire) [inline iroh-unixfs/src/builder.rs:455]: should be the same as the default for `File`

File: iroh-api/src/lib.rs [hunk 6]
```diff
@@ -1,4 +1,5 @@
 mod api;
+mod block_store;
```
Comment (dignifiedquire) [inline iroh-api/src/lib.rs:2]: call this just `store` please

File: iroh-api/benches/add.rs (hunk 1 above)
Comment (ramfox) [reply iroh-api/benches/add.rs:71]: yea it was a mocks issue, I'm side stepping this by not cloning the `Api` struct directly, and cloning the `Resolver` and `Client` instead... should be refactored when we fix the mocks situation.

File: iroh-api/src/block_store.rs (hunk 4 above)
Comment (ramfox) [reply iroh-api/src/block_store.rs:89]: I think it was just being used to compare add speeds. It's a copy-paste from `iroh-unixfs::builder`. I'm comfortable removing it.

File: iroh-unixfs/src/builder.rs (hunk 5 above)
Comment (ramfox) [reply iroh-unixfs/src/builder.rs:455]: ahh, yea, it should also be able to take a `degree` as well, just like `FileBuilder`

thanks

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**